{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded in db!\n",
      "Loaded in transcript sequences!\n"
     ]
    }
   ],
   "source": [
    "# Generates the test data (splits the loading in transcript part with the actual iteration over exon part)\n",
    "\"\"\"\n",
    "tests (not 1 exon, gene level 1)\n",
    "exact transcript to cDNA mapping\n",
    "\n",
    "test_stricter_human_genome/ (not 1 exon, gene level 1, transcript level 1, transcript support level 1)\n",
    "\n",
    "\n",
    "test_stricter_human_genome_with_slight_extension (not just 1 exon, gene level 1, transcript level 1, transcript support level 1, reference sequence doesn't have masked regions (hard masked NNN or soft masked lowercase letters))\n",
    "\n",
    "extend by 2 nt on either side of reference\n",
    "\n",
    "test_stricter_human_genome_with_1_nt_ext (not just 1 exon, gene level 1, transcript level 1, transcript support level 1, reference sequence doesn't have masked regions (hard masked NNN or soft masked lowercase letters))\n",
    "\n",
    "extend reference by 1nt only (to check if pairagon still works)\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "import gffutils\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "def reverse_complement(dna:str):\n",
    "  ans = \"\"\n",
    "  nucleotide_map = {'A':'T','C':'G','G':'C','T':'A'}\n",
    "  for c in dna[-1::-1]:\n",
    "    ans+=nucleotide_map[c]\n",
    "  return ans\n",
    "    \n",
    "    \n",
    "\n",
    "def make_path(path):\n",
    "  if not os.path.exists(path):\n",
    "    os.makedirs(path)\n",
    "\n",
    "annotation_file_name = \"mouse/gencode.vM34.annotation.gff3\"\n",
    "transcript_file_name = \"mouse/gencode.vM34.transcripts.fa\"\n",
    "genome_file_name = \"mouse/GRCm39.primary_assembly.genome.fa\"\n",
    "\n",
    "# db = gffutils.create_db(annotation_file_name, dbfn='mouse.db', force=True, keep_order=False,\n",
    "# merge_strategy='merge', sort_attribute_values=False)\n",
    "\n",
    "db = gffutils.FeatureDB('./mouse/mouse.db', keep_order=True)\n",
    "print(\"Loaded in db!\")\n",
    "transcript_sequences = SeqIO.parse(open(transcript_file_name),'fasta')\n",
    "transcript_id_to_sequence = {seq_record.id.split(\"|\")[0]:seq_record.seq for seq_record in SeqIO.parse(transcript_file_name,\"fasta\") }\n",
    "chromosome_id_to_sequence = {seq_record.id.split(\" \")[0]:seq_record.seq for seq_record in SeqIO.parse(genome_file_name,\"fasta\") }\n",
    "print(\"Loaded in transcript sequences!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/work/cb-proj/path/to/venv/lib/python3.11/site-packages/Bio/SeqRecord.py:229: BiopythonDeprecationWarning: Using a string as the sequence is deprecated and will raise a TypeError in future. It has been converted to a Seq object.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "583\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "\n",
    "sub_dir = \"mouse_mutated_sub\"\n",
    "# for reference genome\n",
    "left_offset = 2\n",
    "right_offset = 2\n",
    "\n",
    "cnt = 0\n",
    "\n",
    "def mutatecDNA(cDNA:str):\n",
    "  mutation_rate = 0.0042\n",
    "  to_sub = {'A':['G','G','C','T'],'T':['G','G','A','C'],'C':['G','G','A','T'],'G':['T']}\n",
    "  mutated_cdna = \"\"\n",
    "  for char in cDNA:\n",
    "    if random.uniform(0,1)<=mutation_rate:\n",
    "      mutated_cdna+=random.choice(to_sub[char])\n",
    "    else:\n",
    "      mutated_cdna+=char\n",
    "  return mutated_cdna\n",
    "\n",
    "for gene in db.features_of_type('gene',order_by='start'):\n",
    "  if gene['level'][0] != \"1\":\n",
    "    continue\n",
    "  \n",
    "  for transcript in db.children(gene,featuretype='transcript',order_by='start'):\n",
    "    transcript_id = transcript.id\n",
    "    cur_chromsome = chromosome_id_to_sequence[transcript.seqid]\n",
    "    if transcript.start is None or transcript.end is None:\n",
    "      continue\n",
    "    if transcript_id in transcript_id_to_sequence:\n",
    "      \n",
    "      fakeCDNA = \"\" # we'll just make this as reference to compare against actual transcript... just in case (but we'll still use the actual transcript)\n",
    "      \n",
    "      \n",
    "      reference_cDNA = transcript_id_to_sequence[transcript_id]\n",
    "      if transcript.start-left_offset<1 or transcript.end+right_offset>len(cur_chromsome):\n",
    "        print(\"Transcript starts too early! Pairagon will be upset\")\n",
    "        continue\n",
    "      reference_sequence = cur_chromsome[transcript.start-1-left_offset:transcript.end+right_offset]\n",
    "      exon_count = 0\n",
    "      exon_intervals = []\n",
    "      # if not transcript['level'][0]==\"1\":\n",
    "      # print(transcript['levefl'][0] != \"1\" , transcript['transcript_support_level'][0]!=\"1\")\n",
    "      if 'level' not in transcript.attributes or transcript['level'][0] != \"1\" or 'transcript_support_level' not in transcript.attributes or transcript['transcript_support_level'][0]!=\"1\":\n",
    "        continue\n",
    "      # if 'MANE_Select' not in transcript.attributes:\n",
    "      #   continue\n",
    "      \n",
    "      # if 'ccdsid' not in transcript.attributes or len(transcript['ccdsid']) == 0:\n",
    "      #   # print(\"skipping non-coding transcript\")\n",
    "      #   continue\n",
    "      if not set(reference_sequence) <= set(\"ACTG\"):\n",
    "          print(\"Masked region detected, skipping\",transcript_id,reference_sequence)\n",
    "          # no N's or lowercase letters allowed (basically we don't want soft-masking/uncertain data)\n",
    "          continue\n",
    "      \n",
    "      for exon in db.children(transcript,featuretype='exon',order_by='start'):\n",
    "        assert(exon['level'][0]==\"1\")\n",
    "        assert(exon[\"transcript_support_level\"][0]==\"1\")\n",
    "        \n",
    "        assert(exon.start is not None and exon.end is not None)\n",
    "        startI = exon.start-transcript.start\n",
    "        endI = exon.end-transcript.start\n",
    "        fakeCDNA+=reference_sequence[startI+left_offset:endI+1+left_offset] #also offset here. this is a bit annoying\n",
    "        exon_count+=1\n",
    "        exon_intervals.append(f\"{startI+1+left_offset}/{endI+1+left_offset}\") # return it back to 1-indexing (and then add the 2-nt offset at the start)\n",
    "      if exon_count <= 1:\n",
    "        print(\"1 exon\",transcript_id)\n",
    "        assert(exon_count!= 0)\n",
    "        continue # really boring transcript\n",
    "      # print(fakeCDNA)\n",
    "      # print(reference_cDNA)\n",
    "      assert(fakeCDNA==reference_cDNA or fakeCDNA==reverse_complement(reference_cDNA)) \n",
    "      \n",
    "      exon_intervals.append(\"+\" if fakeCDNA==reference_cDNA else \"-\") # directionality (template vs. non-template strand)\n",
    "      reference_cDNA = mutatecDNA(reference_cDNA) # da only substitution mutation\n",
    "      assert(gene.id != None)\n",
    "      make_path(f\"{sub_dir}/{transcript_id}\")\n",
    "      SeqIO.write(SeqRecord(reference_sequence,transcript_id,description=gene.id),tfn:=f\"{sub_dir}/{transcript_id}/genome.fa\",\"fasta\")\n",
    "      SeqIO.write(SeqRecord(reference_cDNA,f\"cDNA_{transcript_id}\",description=gene.id),cdna_fn:=f\"{sub_dir}/{transcript_id}/cdna.fa\",\"fasta\") # type: ignore\n",
    "      open(f\"{sub_dir}/{transcript_id}/ans.txt\", \"w\").write(\"\\n\".join(exon_intervals))\n",
    "      cnt += 1\n",
    "      \n",
    "      # subprocess.run(f\"/Users/work/Documents/SD_Projects/gmap-2024/bin/gmap {cdna_fn} -g {tfn} -E genomic -f 3 > tests/{transcript_id}/result.gmap\",check=True,shell=True,stdout=subprocess.DEVNULL,stderr=subprocess.DEVNULL)\n",
    "    else:\n",
    "      print(f\"Skipping {transcript_id} cuz not found\")\n",
    "\n",
    "print(cnt)\n",
    "# This time filtering out everything where genome is masked\n",
    "# Filter out non-coding regions"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
