{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'mgi_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'protein_id', 'ccdsid', 'ont']\n",
      "/tmp/ipykernel_70497/1506973251.py:24: DeprecationWarning: `group_by` iteration will change to always return group identifiers as tuples. Pass `by` as a list to silence this warning, e.g. `group_by(['gene_id'])`.\n",
      "  gene_groups = {name: group for name, group in exon_details.group_by(\"gene_id\")}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully Loaded Annotated Genome\n"
     ]
    }
   ],
   "source": [
    "import gtfparse\n",
    "import polars as pl\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "def loadingGenome(): # returns gene_groups\n",
    "    # Load the GTF file\n",
    "    gtf_file = 'gencode.vM34.chr_patch_hapl_scaff.basic.annotation.gtf'\n",
    "    gtf_data = gtfparse.read_gtf(gtf_file)\n",
    "\n",
    "    # Filter to get only exons using Polars\n",
    "    exons = gtf_data.filter(gtf_data['feature'] == 'exon')\n",
    "\n",
    "    # Example to display exon positions and gene name using Polars\n",
    "    exon_details = exons.select(['seqname', 'start', 'end', 'feature', 'gene_name', 'gene_id'])\n",
    "    exon_details = exon_details.with_columns((exons['end'] - exons['start'] + 1).alias('exon_length'))\n",
    "    exon_details = exon_details.sort(['gene_name', 'start'])\n",
    "\n",
    "\n",
    "    # sort exons by gene\n",
    "    gene_groups = {name: group for name, group in exon_details.group_by(\"gene_id\")}\n",
    "\n",
    "    print(\"Successfully Loaded Annotated Genome\")\n",
    "    return gene_groups\n",
    "  \n",
    "gene_groups = loadingGenome()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###################\n",
    "def createTranscripts(gene_groups):\n",
    "    # Use exist_ok to avoid error if the directory already exists\n",
    "    os.makedirs('simulatedmRNAs/', exist_ok=True)\n",
    "\n",
    "    # load the sequences\n",
    "    sequences = load_fasta()\n",
    "\n",
    "    # loop through every gene in gene_groups\n",
    "    for gene in list(gene_groups.keys()):\n",
    "\n",
    "        # make a folder to store the mRNA strings for this gene\n",
    "        loc = os.path.join('simulatedmRNAs', gene)  # Use os.path.join for OS-independent path handling\n",
    "        os.makedirs(loc, exist_ok=True)  # Use exist_ok here as well\n",
    "\n",
    "        # get the number of exons for this gene\n",
    "        exon_count = gene_groups[gene].shape[0]\n",
    "\n",
    "        # randomly choose how many exons will be present in this transcript\n",
    "        number_of_exons = exon_count\n",
    "        # number_of_exons = random.randint(1, exon_count)\n",
    "\n",
    "        # randomly choose a subset of size number_of_exons\n",
    "        exon_subset_indices = random.sample(range(exon_count), number_of_exons)\n",
    "        exon_subset_indices.sort() # make sure the order is sorted\n",
    "\n",
    "        # initialize exon_boundaries\n",
    "        exon_boundaries = [] # list of tuples\n",
    "        \n",
    "        # store this in our folder\n",
    "        \n",
    "        # file_path = os.path.join(loc, 'ans.txt')\n",
    "        # with open(file_path, 'w') as file:  # Using 'with' ensures the file is closed properly after its suite finishes\n",
    "        #     for exon_idx in exon_subset_indices: \n",
    "        #         start, end = gene_groups[gene]['start'][exon_idx], gene_groups[gene]['end'][exon_idx]\n",
    "\n",
    "        #         assert(type(start) == type(end) == int)\n",
    "\n",
    "        #         exon_boundaries.append((start, end))\n",
    "        #         file.write(str(start) + '/' + str(end) + '\\n')\n",
    "        #     file.write('-')\n",
    "\n",
    "        # create file to know what sequence this gene comes from (to know what to align it to)\n",
    "        # file_path = os.path.join(loc, 'sequence_ref.txt')\n",
    "        # with open(file_path, 'w') as file:\n",
    "        #     file.write(gene_groups[gene]['seqname'][0])\n",
    "\n",
    "\n",
    "\n",
    "        # file_path = os.path.join(loc, 'cdna.fa')\n",
    "        # with open(file_path, 'w') as file:  # Using 'with' ensures the file is closed properly after its suite finishes\n",
    "        #     file.write('>' + 'cDNA_' + gene + '\\n')\n",
    "        #     sequence = sequences.get(gene_groups[gene][\"seqname\"][0])\n",
    "\n",
    "        #     if sequence: \n",
    "\n",
    "        #         # ensuring the data we are adding is actually being added\n",
    "        #         current_length = 0\n",
    "        #         expected_length = 0\n",
    "\n",
    "        #         for start, end in exon_boundaries: \n",
    "        #             assert(type(start) == type(end) == int)\n",
    "\n",
    "        #             expected_length += end - start + 1\n",
    "\n",
    "        #             if (start >= len(sequence)): \n",
    "        #                 print(\"starting index is greater than length of sequence\")\n",
    "        #                 continue\n",
    "                    \n",
    "        #             if (end >= len(sequence)):\n",
    "        #                 print(\"ending index is greater than the length of sequence\")\n",
    "        #                 continue\n",
    "                        \n",
    "        #             # Adjust for Python's 0-based index; subtract 1 from start\n",
    "        #             exon_seq = sequence[start-1:end]\n",
    "\n",
    "        #             current_length += len(exon_seq) \n",
    "\n",
    "                    \n",
    "        #             file.write(str(exon_seq))\n",
    "        #     else: \n",
    "        #         print(\"Sequence type is NONE\")\n",
    "        #         continue\n",
    "            \n",
    "        #     if (expected_length != current_length):\n",
    "        #         print(\"Expected Length: \" + str(expected_length), \"Current Length: \" + str(current_length))\n",
    "\n",
    "        #     file.write('\\n')\n",
    "\n",
    "    print(\"Successfully Created Artificial cDNA strands\")\n",
    "############################################################\n",
    "createTranscripts(gene_groups)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file = 'gencode.vM34.chr_patch_hapl_scaff.basic.annotation.gtf'\n",
    "gtf_data = gtfparse.read_gtf(gtf_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
